### ISSUE TITLE
Test worker_steal_count hangs sometimes

### ISSUE THREAD

**Issue by Darksonn** [tokio-rs/tokio#6470]
I tried fixing this in the past with #6327, but it seems like that attempt was not enough to fix it, because it's still happening.

https://github.com/tokio-rs/tokio/blob/431b7c52636850820dd1f2b45e02c06cf366cb1c/tokio/tests/rt_metrics.rs#L177-L213

For now, I've ignored the test in #6471, but it should be fixed.

**Comment by ineol** [comment]
I couldn't figure it out, but I have a few observations that could hopefully be useful.

1. The usual cause of non-termination is that the inner task is never scheduled, and one worker gets parked nonethelesss
2. If one adds a call to `std::thread::sleep` in `runtime::scheduler::multi_thread::worker::Context::run` after trying to sleep and before parking. ([this line](https://github.com/tokio-rs/tokio/blob/6fcd9c02176bf3cd570bc7de88edaa3b95ea480a/tokio/src/runtime/scheduler/multi_thread/worker.rs#L541))
3. Always notifying in this function called just before parking a core seems to improve the situation: without the `thread::sleep` mentioned above, I cannot reproduce the bug anymore (in 10^6 runs)
  https://github.com/tokio-rs/tokio/blob/6fcd9c02176bf3cd570bc7de88edaa3b95ea480a/tokio/src/runtime/scheduler/multi_thread/worker.rs#L898-L903
With this change, the send line 196 of the test runs, and the receive happens in the outer task, but it still not terminate sometimes, which confused me.

It could be that somehow, 
1. while worker 1 is searching for work in the global queue, worker 2 does not transition to searching because of [this condition](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/scheduler/multi_thread/idle.rs#L106), then it fails to find work. 
2. then, worker 1 runs the task and enqueues a task to its local queue, and gets stuck at `rx.recv()`
3. then worker 2  parks, the `transition_to_parked` runs, and there are zero workers in a searching state, and so it doesn't try to notify, and it gets parked
4. the final result is that one worker is executing the task, waiting on the channel, and the other worker is parked.

Instead of always trying to notify, changing the condition below with `(is_searching as usize) == prev.num_searching()` has the same effect the change in point 3 above.

https://github.com/tokio-rs/tokio/blob/6fcd9c02176bf3cd570bc7de88edaa3b95ea480a/tokio/src/runtime/scheduler/multi_thread/idle.rs#L188-L198

**Comment by jofas** [comment]
Isn't this the same problematic blocking of the runtime that was observed in #6847 and could be fixed by applying the fix in #6916 to this test?

**Comment by Darksonn** [comment]
That sounds likely!
